src/OpenFeature.Contrib.Providers.Flagd/Cache.cs

```diff
@@ -0,0 +1,170 @@
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
+
+namespace OpenFeature.Contrib.Providers.Flagd
+{
+    internal interface ICache<TKey, TValue>
+    {
+        void Add(TKey key, TValue value);
+        TValue TryGet(TKey key);
+        void Delete(TKey key);
+        void Purge();
+    }
+    class LRUCache<TKey, TValue> : ICache<TKey, TValue> where TValue : class
+    {
+        private readonly int _capacity;
+        private readonly Dictionary<TKey, Node> _map;
+        private Node _head;
+        private Node _tail;
+
+        private System.Threading.Mutex _mtx;
+
+        public LRUCache(int capacity)
+        {
+            _capacity = capacity;
+            _map = new Dictionary<TKey, Node>();
+            _mtx = new System.Threading.Mutex();
+        }
+
+        public TValue TryGet(TKey key)
+        {
+            try
+            {
+                _mtx.WaitOne();
+                if (_map.TryGetValue(key, out Node node))
+                {
+                    MoveToFront(node);
+                    return node.Value;
+                }
+                return default(TValue);
+            }
+            finally
+            {
+                _mtx.ReleaseMutex();
+            }
```

Comment:
You may want to consider wrapping your mutex is an IDisposable interface, which would allow you to remove all these `finally` blocks. Here's a blog post explaining it: https://www.twilio.com/blog/c-8-making-use-of-using-declarations

Reply:
that's a great suggestion, thanks! I was not too happy about the `finally` blocks either, especially being used to Go's `defer`. The approach with the `IDisposable` interface will be much closer to what I'm used to with Go, so I'll be happy to look into that :)